src/memray/_memray/inject.cpp

```diff
@@ -0,0 +1,245 @@
+#define Py_LIMITED_API 0x03070000
+#include "Python.h"
+
+#include <netdb.h>
+#include <netinet/in.h>
+#include <netinet/tcp.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+
+#include <iostream>
+
+namespace memray {
+namespace {  // unnamed
+
+int
+connect_client(int port)
+{
+    struct addrinfo hints = {};
+    struct addrinfo* all_addresses = nullptr;
+    struct addrinfo* curr_address = nullptr;
+    int rv;
+    int sockfd;
+
+    hints.ai_family = AF_UNSPEC;
+    hints.ai_socktype = SOCK_STREAM;
+
+    std::string port_str = std::to_string(port);
+    if ((rv = ::getaddrinfo(nullptr, port_str.c_str(), &hints, &all_addresses)) != 0) {
+        std::cerr << "getaddrinfo() failed while trying to attach Memray: " << ::gai_strerror(rv);
+        return -1;
+    }
+
+    for (curr_address = all_addresses; curr_address != nullptr; curr_address = curr_address->ai_next) {
+        sockfd = ::socket(curr_address->ai_family, curr_address->ai_socktype, curr_address->ai_protocol);
+        if (sockfd == -1) {
+            continue;
+        }
+
+        if (::connect(sockfd, curr_address->ai_addr, curr_address->ai_addrlen) == -1) {
+            ::close(sockfd);
+            continue;
+        }
+        break;
+    }
+    ::freeaddrinfo(all_addresses);
+    if (curr_address == nullptr) {
+        return -1;
+    }
+    return sockfd;
+}
+
+bool
+sendall(int fd, std::string_view data)
+{
+    size_t length = data.size();
+    while (length) {
+        ssize_t ret = ::send(fd, &data[0], length, 0);
+        if (ret < 0 && errno != EINTR) {
+            return false;
+        } else if (ret >= 0) {
+            data.remove_prefix(ret);
+            length -= ret;
+        }
+    }
+    return true;
+}
+
+bool
+recvall(int fd, std::string* data)
+{
+    data->clear();
+
+    char buf[4096];
+    while (true) {
+        ssize_t ret = ::recv(fd, buf, sizeof(buf), 0);
+        if (ret < 0 && errno != EINTR) {
+            return false;
+        } else if (ret > 0) {
+            *data += std::string_view(buf, ret);
+        } else if (ret == 0) {
+            return true;
+        }
+    }
+}
+
+// Clear the error indicator, return a string representing the error.
+std::string
+Memray_PyErr_ToString()
+{
+    std::string ret;
+
+    if (!PyErr_Occurred()) {
+        return ret;
+    }
+
+    PyObject* type;
+    PyObject* val;
+    PyObject* tb;
+    PyErr_Fetch(&type, &val, &tb);
+    PyErr_NormalizeException(&type, &val, &tb);
+
+    PyObject* exc_repr = PyObject_Repr(val);
+    if (!exc_repr) {
+        PyErr_Clear();
+        ret = "unknown exception (`repr(exc)` failed)!";
+    } else {
+        PyObject* utf8 = PyUnicode_AsUTF8String(exc_repr);
+        if (!utf8) {
+            PyErr_Clear();
+            ret = "unknown exception (`repr(exc).encode('utf-8')` failed)!";
+        } else {
+            ret = PyBytes_AsString(utf8);
```

Comment:
Very small nit (feel free to ignore): This can potentially return NULL, although is only if the object is not a bytes object. If we want to skip the error checking, let's add an assert with a `PyBytesCheckExact` before OR an assert that the return is not NULL.

Reply:
I suppose I can add an assert, but - is there some case where `PyUnicode_AsUTF8String` returns something other than `NULL` or a `bytes` object? I didn't think that was possible/within contract.

Reply:
Discussed offline: there's no other case that would be within contract, so I'm not going to check this. We're not in the business of testing that CPython itself is behaving correctly.